File: 003/problem_003.go
package main

import "fmt"
import "math"

func IsPrime(n int64) bool{
    s := int64(math.Sqrt(float64(n)));

	for i := s; i > 1; i-- {
		if(n%i == 0) {
			return false;
		}
	}

	return true;
}

func MaxPrimeFactor(n int64) int64 {
    s := int64(math.Sqrt(float64(n)));

    for i :=s; i > 1; i-- {
		if(n%i == 0 && IsPrime(i)) {
			return i;
		}
	}

	return 1;
}

func main() {
	fmt.Println("result = ",MaxPrimeFactor(600851475143));
}

File: 003/problem_003.cpp
#include <iostream>
#include <math.h>
#include <assert.h>
#include <stdint.h>

using namespace std;

bool isPrime(int n)
{
    assert(n > 0);

    int s = sqrt(n);

    for(int i=s; i > 1; --i)
    {
        if(n%i == 0)
        {
            return false;
        }
    }

    return true;
}

int MaxPrimeFactor(uint64_t n)
{
    assert(n > 0);
    int s = sqrt(n);

    for(int i=s; i > 1; --i)
    {
        if(n%i == 0 && isPrime(i))
        {
            return i;
        }
    }

    return 1;
}

int main()
{
    cout<<"result = "<<MaxPrimeFactor(600851475143)<<endl;
    return 0;
}
